### PR TITLE
Replace is_callable checks with type hints

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #14330 |
| License | MIT |
| Doc PR | - |

Also removes tests checking the exceptions thrown from
the removed is_callable checks.

### CODE REVIEW

**Comment by Tobion** [comment]
This is the same BC break as in ##14330. You get a warning for overwritten method that the signature is not the same: https://3v4l.org/ruPcK

But I'm :+1: to add callable typehints.

**Comment by stof** [comment]
@Tobion in this PR, it is OK:

- constructors don't need to have a matching signature
- most other classes are not meant to be extended (the only case I see is ``Command``, and you won't override ``setCode`` in such case but ``execute``)

**Comment by stof** [comment]
@nicolas-grekas ``HttpKernelTest::testHandleWhenTheControllerIsNotACallable`` is broken here

and it would be great to fix the deps=low tests of the Translation component

**Comment by nicolas-grekas** [comment]
I added more callable type hints that should all qualify as constructor's or methods not meant to be extended.
@stof I'll look at deps=low tests if you look at #16097 ;-)

File: src/Symfony/Component/VarDumper/Exception/ThrowingCasterException.php [hunk 1]
```diff
@@ -17,10 +17,9 @@
 class ThrowingCasterException extends \Exception
 {
     /**
-     * @param callable   $caster The failing caster
-     * @param \Exception $prev   The exception thrown from the caster
+     * @param \Exception $prev The exception thrown from the caster
      */
-    public function __construct($caster, \Exception $prev)
+    public function __construct(\Exception $prev)
```
Comment (stof) [inline src/Symfony/Component/VarDumper/Exception/ThrowingCasterException.php]: shouldn't this be changed in older branches actually ?

Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Exception/ThrowingCasterException.php]: Yep, see #16129

File: src/Symfony/Component/Translation/PluralizationRules.php [hunk 2]
```diff
@@ -189,12 +189,10 @@ public static function get($number, $locale)
     /**
      * Overrides the default plural rule for a given locale.
      *
-     * @param string $rule   A PHP callable
-     * @param string $locale The locale
-     *
-     * @throws \LogicException
+     * @param callable $rule   A PHP callable
```
Comment (stof) [inline src/Symfony/Component/Translation/PluralizationRules.php]: this phpdoc fix should actually be applied in older branches too

**Comment by Tobion** [comment]
:+1: 

Status: Reviewed

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**